src/Build/Definition/Project.cs

```diff
@@ -2087,6 +2093,27 @@ public override bool IsDirty
                 }
             }
 
+            public bool GlobalPropertiesContains(string key)
+            {
+                return _data.GlobalPropertiesDictionary.Contains(key);
+            }
+
+            public int GlobalPropertiesCount()
+            {
+                return _data.GlobalPropertiesDictionary.Count;
+            }
+
+            public IEnumerable<KeyValuePair<string, string>> GlobalPropertiesEnumerable()
```

Comment:
It looks like GlobalPropertiesContains/Count will cut the overhead to get a new copy of the dictionary, but I am not sure about GlobalPropertiesEnumerable, that seems to make another copy anyway?  Maybe copying to a list is faster than creating a dictionary?  If our goal is to improve on that, maybe you can create an array, and fill it, instead of using a list.

Reply:
I modified the Enumerable to use yield return. The uses don't seem like they need a fully instantiated object to me, though please do correct me if I'm wrong 🙂

I also pushed the switch to virtual methods. Does that look good to you?